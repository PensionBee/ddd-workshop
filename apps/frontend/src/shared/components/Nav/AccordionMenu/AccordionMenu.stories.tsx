import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { GridContainer, GridItem, GridRow } from "@shared/components/Grid";
import { LIST_WITH_MIXED_ITEMS } from "../Nav.content";
import type { AccordionMenuProps } from "./AccordionMenu";
import { AccordionMenu } from "./AccordionMenu";

export default {
  title: "Shared/Menus/Accordion Menu",
  component: AccordionMenu,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof AccordionMenu>;

const AccordionMenuTemplate: ComponentStory<typeof AccordionMenu> = (
  args: AccordionMenuProps
) => {
  return (
    <GridContainer>
      <GridRow>
        <GridItem>
          <AccordionMenu {...args} />
        </GridItem>
      </GridRow>
    </GridContainer>
  );
};

export const Default = AccordionMenuTemplate.bind({});
Default.args = {
  list: LIST_WITH_MIXED_ITEMS,
};
