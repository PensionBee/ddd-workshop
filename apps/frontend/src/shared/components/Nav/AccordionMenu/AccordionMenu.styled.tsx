import styled from "styled-components";
import { mediaFrom } from "@shared/styling/mediaQueries";
import { LargeSystemIcon } from "../../Svg/Svg";
import { theme } from "@shared/styling/theme";

export const StyledAccordionButton = styled.button`
  width: 100%;
  cursor: pointer;
  border-radius: 4px;
  &:focus-visible {
    outline: 3px solid ${() => theme.colors.blue100};
  }
`;

export const StyledMenuList = styled.ul`
  width: 100%;
`;

export const StyledCaret = styled(LargeSystemIcon)`
  position: relative;
  top: 1;
  transform: rotate(-180deg);
  transition: transform 250ms ease;
  [data-state="open"] & {
    transform: rotate(0deg);
  }
`;

export const StyledMenuItem = styled.li`
  width: 100%;
  > a,
  > button {
    display: flex;
    align-items: center;
    justify-content: space-between;
    column-gap: 10px;
    padding-top: 14px;
    padding-bottom: 14px;
    border-bottom: 1px solid ${() => theme.colors.grey4};
  }
  ul {
    ${mediaFrom.md} {
      padding-left: 40px;
    }
  }
`;
