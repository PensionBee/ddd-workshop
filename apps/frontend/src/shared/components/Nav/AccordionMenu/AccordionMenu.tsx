import React, { useState } from "react";
import {
  StyledAccordionButton,
  StyledMenuList,
  StyledCaret,
} from "./AccordionMenu.styled";
import type { NavMenuItem } from "../Nav.types";
import { Box } from "../../Box/Box";
import { AccordionMenuItem } from "./AccordionMenuItem/AccordionMenuItem";

export type AccordionMenuProps = {
  list: NavMenuItem[];
};

export const AccordionMenu: React.FC<AccordionMenuProps> = ({
  list,
  ...props
}) => {
  const [menuOpen, setMenuOpen] = useState(false);

  const handleToggleMenuOpen = () => {
    return setMenuOpen((menuOpen) => !menuOpen);
  };

  return (
    <Box
      d="flex"
      w="100%"
      justify="center"
      align="center"
      data-rc="AccordionMenu"
      {...props}
    >
      <StyledAccordionButton
        onClick={handleToggleMenuOpen}
        data-state={menuOpen ? "open" : "closed"}
      >
        <Box d="flex" align="center" justify="center" colGap={10} py={14}>
          Menu
          <StyledCaret icon="ChevronYellow" aria-hidden />
        </Box>
      </StyledAccordionButton>
      <Box borderBottom={1} borderColor="grey4" w="100%" />
      {menuOpen ? (
        <StyledMenuList>
          {list.map((item) => (
            <AccordionMenuItem key={item.children} {...item} />
          ))}
        </StyledMenuList>
      ) : null}
    </Box>
  );
};
