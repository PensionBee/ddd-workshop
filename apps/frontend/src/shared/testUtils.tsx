import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
