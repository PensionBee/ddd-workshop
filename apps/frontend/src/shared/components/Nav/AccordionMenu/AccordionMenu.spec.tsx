import { fireEvent, render } from "@shared/testUtils";
import {
  LIST,
  LIST_WITH_HIDDEN_LINK,
  LIST_WITH_MIXED_ITEMS,
} from "../Nav.content";
import { AccordionMenu } from "./AccordionMenu";

const MENU_TEXT = "Menu";

describe("AccordionMenuItem", () => {
  it("should render correctly with list", () => {
    const { getByText, asFragment } = render(<AccordionMenu list={LIST} />);

    expect(getByText(MENU_TEXT)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not render if hidden", () => {
    const { queryByText, asFragment } = render(
      <AccordionMenu list={LIST_WITH_HIDDEN_LINK} />
    );

    expect(queryByText(LIST_WITH_HIDDEN_LINK[0].children)).toBe(null);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with data attributes", () => {
    const TEST_ATTRIBUTE = "test-attribute";
    const { asFragment } = render(
      <AccordionMenu list={LIST} data-test={TEST_ATTRIBUTE} />
    );

    const item = document.querySelector(`[data-test="${TEST_ATTRIBUTE}"]`);

    expect(item).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should toggle the list on click", () => {
    const { queryByText, getByText, asFragment } = render(
      <AccordionMenu list={LIST} />
    );
    const toggleButton = getByText(MENU_TEXT);

    LIST.forEach((item) => expect(queryByText(item.children)).toBe(null));
    fireEvent.click(toggleButton);
    LIST.forEach((item) => expect(queryByText(item.children)).toBeVisible());
    fireEvent.click(toggleButton);
    LIST.forEach((item) => expect(queryByText(item.children)).toBe(null));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not render hidden sub list items", () => {
    const { getByText, queryByText, asFragment } = render(
      <AccordionMenu list={LIST_WITH_MIXED_ITEMS} />
    );
    const toggleButton = getByText(MENU_TEXT);

    fireEvent.click(toggleButton);
    LIST_WITH_MIXED_ITEMS.forEach((item) => {
      if (item.hidden) {
        expect(queryByText(item.children)).toBe(null);
      } else {
        expect(queryByText(item.children)).toBeVisible();
      }
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
