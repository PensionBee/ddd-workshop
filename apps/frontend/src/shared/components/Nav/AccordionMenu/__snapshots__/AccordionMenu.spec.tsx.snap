// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccordionMenuItem > should not render hidden sub list items 1`] = `
<DocumentFragment>
  <div
    class="sc-jSMdHm gRtYZt"
    data-rc="AccordionMenu"
  >
    <button
      class="sc-gsnScm eGwlpA"
      data-state="open"
    >
      <div
        class="sc-jSMdHm FxIeT"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bczSft cMucIl sc-hKMtDE kFnGoN"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg
            height="7"
            viewBox="0 0 12 7"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.406.252l-5.16 5.28a.874.874 0 0 0 0 1.216.827.827 0 0 0 1.189 0L6 2.076l4.565 4.672a.827.827 0 0 0 1.189 0 .874.874 0 0 0 0-1.216L6.594.252a.828.828 0 0 0-1.188 0z"
              fill="#F9C000"
              fill-rule="nonzero"
            />
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSMdHm eQuTnj"
    />
    <ul
      class="sc-dkzCjb fhmfsU"
    >
      <li
        class="sc-eCYdKt fhdMkn"
        data-rc="AccordionMenuItem"
      >
        <a
          class="sc-gKXNOb jSssGy"
          data-rc="Link"
          href="/"
          target="_self"
        >
          <div
            class="sc-jSMdHm fOjiZJ"
          >
            Visible Item
          </div>
        </a>
      </li>
      <li
        class="sc-eCYdKt fhdMkn"
        data-rc="AccordionMenuItem"
      >
        <button
          class="sc-gsnScm eGwlpA"
          data-state="closed"
        >
          SubList Item
          <span
            aria-hidden="true"
            class="sc-bczSft cMucIl sc-hKMtDE kFnGoN"
            data-rc="LargeSystemIcon.ChevronYellow"
          >
            <svg
              height="7"
              viewBox="0 0 12 7"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.406.252l-5.16 5.28a.874.874 0 0 0 0 1.216.827.827 0 0 0 1.189 0L6 2.076l4.565 4.672a.827.827 0 0 0 1.189 0 .874.874 0 0 0 0-1.216L6.594.252a.828.828 0 0 0-1.188 0z"
                fill="#F9C000"
                fill-rule="nonzero"
              />
            </svg>
          </span>
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`AccordionMenuItem > should not render if hidden 1`] = `
<DocumentFragment>
  <div
    class="sc-jSMdHm gRtYZt"
    data-rc="AccordionMenu"
  >
    <button
      class="sc-gsnScm eGwlpA"
      data-state="closed"
    >
      <div
        class="sc-jSMdHm FxIeT"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bczSft cMucIl sc-hKMtDE kFnGoN"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg
            height="7"
            viewBox="0 0 12 7"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.406.252l-5.16 5.28a.874.874 0 0 0 0 1.216.827.827 0 0 0 1.189 0L6 2.076l4.565 4.672a.827.827 0 0 0 1.189 0 .874.874 0 0 0 0-1.216L6.594.252a.828.828 0 0 0-1.188 0z"
              fill="#F9C000"
              fill-rule="nonzero"
            />
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSMdHm eQuTnj"
    />
  </div>
</DocumentFragment>
`;

exports[`AccordionMenuItem > should render correctly with list 1`] = `
<DocumentFragment>
  <div
    class="sc-jSMdHm gRtYZt"
    data-rc="AccordionMenu"
  >
    <button
      class="sc-gsnScm eGwlpA"
      data-state="closed"
    >
      <div
        class="sc-jSMdHm FxIeT"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bczSft cMucIl sc-hKMtDE kFnGoN"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg
            height="7"
            viewBox="0 0 12 7"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.406.252l-5.16 5.28a.874.874 0 0 0 0 1.216.827.827 0 0 0 1.189 0L6 2.076l4.565 4.672a.827.827 0 0 0 1.189 0 .874.874 0 0 0 0-1.216L6.594.252a.828.828 0 0 0-1.188 0z"
              fill="#F9C000"
              fill-rule="nonzero"
            />
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSMdHm eQuTnj"
    />
  </div>
</DocumentFragment>
`;

exports[`AccordionMenuItem > should render with data attributes 1`] = `
<DocumentFragment>
  <div
    class="sc-jSMdHm gRtYZt"
    data-rc="AccordionMenu"
    data-test="test-attribute"
  >
    <button
      class="sc-gsnScm eGwlpA"
      data-state="closed"
    >
      <div
        class="sc-jSMdHm FxIeT"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bczSft cMucIl sc-hKMtDE kFnGoN"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg
            height="7"
            viewBox="0 0 12 7"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.406.252l-5.16 5.28a.874.874 0 0 0 0 1.216.827.827 0 0 0 1.189 0L6 2.076l4.565 4.672a.827.827 0 0 0 1.189 0 .874.874 0 0 0 0-1.216L6.594.252a.828.828 0 0 0-1.188 0z"
              fill="#F9C000"
              fill-rule="nonzero"
            />
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSMdHm eQuTnj"
    />
  </div>
</DocumentFragment>
`;

exports[`AccordionMenuItem > should toggle the list on click 1`] = `
<DocumentFragment>
  <div
    class="sc-jSMdHm gRtYZt"
    data-rc="AccordionMenu"
  >
    <button
      class="sc-gsnScm eGwlpA"
      data-state="closed"
    >
      <div
        class="sc-jSMdHm FxIeT"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bczSft cMucIl sc-hKMtDE kFnGoN"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg
            height="7"
            viewBox="0 0 12 7"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.406.252l-5.16 5.28a.874.874 0 0 0 0 1.216.827.827 0 0 0 1.189 0L6 2.076l4.565 4.672a.827.827 0 0 0 1.189 0 .874.874 0 0 0 0-1.216L6.594.252a.828.828 0 0 0-1.188 0z"
              fill="#F9C000"
              fill-rule="nonzero"
            />
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSMdHm eQuTnj"
    />
  </div>
</DocumentFragment>
`;

exports[`AccordionMenuItem should not render hidden sub list items 1`] = `
<DocumentFragment>
  <div
    class="sc-jSUZER cGzYVp"
    data-rc="AccordionMenu"
  >
    <button
      class="sc-gswNZR dmdjWu"
      data-state="open"
    >
      <div
        class="sc-jSUZER ljdiQH"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bcXHqe lmcdLK sc-hLBbgP cqdzXG"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg>
            chevron-yellow.svg
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSUZER fuBeJn"
    />
    <ul
      class="sc-dkrFOg fXYAxm"
    >
      <li
        class="sc-eDvSVe Jedgj"
        data-rc="AccordionMenuItem"
      >
        <a
          class="sc-gKPRtg hDcrGb"
          data-rc="Link"
          href="/"
          target="_self"
        >
          <div
            class="sc-jSUZER iKLYLh"
          >
            Visible Item
          </div>
        </a>
      </li>
      <li
        class="sc-eDvSVe Jedgj"
        data-rc="AccordionMenuItem"
      >
        <button
          class="sc-gswNZR dmdjWu"
          data-state="closed"
        >
          SubList Item
          <span
            aria-hidden="true"
            class="sc-bcXHqe lmcdLK sc-hLBbgP cqdzXG"
            data-rc="LargeSystemIcon.ChevronYellow"
          >
            <svg>
              chevron-yellow.svg
            </svg>
          </span>
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`AccordionMenuItem should not render if hidden 1`] = `
<DocumentFragment>
  <div
    class="sc-jSUZER cGzYVp"
    data-rc="AccordionMenu"
  >
    <button
      class="sc-gswNZR dmdjWu"
      data-state="closed"
    >
      <div
        class="sc-jSUZER ljdiQH"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bcXHqe lmcdLK sc-hLBbgP cqdzXG"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg>
            chevron-yellow.svg
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSUZER fuBeJn"
    />
  </div>
</DocumentFragment>
`;

exports[`AccordionMenuItem should render correctly with list 1`] = `
<DocumentFragment>
  <div
    class="sc-jSUZER cGzYVp"
    data-rc="AccordionMenu"
  >
    <button
      class="sc-gswNZR dmdjWu"
      data-state="closed"
    >
      <div
        class="sc-jSUZER ljdiQH"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bcXHqe lmcdLK sc-hLBbgP cqdzXG"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg>
            chevron-yellow.svg
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSUZER fuBeJn"
    />
  </div>
</DocumentFragment>
`;

exports[`AccordionMenuItem should render with data attributes 1`] = `
<DocumentFragment>
  <div
    class="sc-jSUZER cGzYVp"
    data-rc="AccordionMenu"
    data-test="test-attribute"
  >
    <button
      class="sc-gswNZR dmdjWu"
      data-state="closed"
    >
      <div
        class="sc-jSUZER ljdiQH"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bcXHqe lmcdLK sc-hLBbgP cqdzXG"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg>
            chevron-yellow.svg
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSUZER fuBeJn"
    />
  </div>
</DocumentFragment>
`;

exports[`AccordionMenuItem should toggle the list on click 1`] = `
<DocumentFragment>
  <div
    class="sc-jSUZER cGzYVp"
    data-rc="AccordionMenu"
  >
    <button
      class="sc-gswNZR dmdjWu"
      data-state="closed"
    >
      <div
        class="sc-jSUZER ljdiQH"
      >
        Menu
        <span
          aria-hidden="true"
          class="sc-bcXHqe lmcdLK sc-hLBbgP cqdzXG"
          data-rc="LargeSystemIcon.ChevronYellow"
        >
          <svg>
            chevron-yellow.svg
          </svg>
        </span>
      </div>
    </button>
    <div
      class="sc-jSUZER fuBeJn"
    />
  </div>
</DocumentFragment>
`;
